django/contrib/auth/hashers.py

```diff
@@ -228,7 +231,11 @@ def safe_summary(self, encoded):
         raise NotImplementedError('subclasses of BasePasswordHasher must provide a safe_summary() method')
 
     def must_update(self, encoded):
-        return False
+        try:
+            # A string of length: N / log_2(26+26+10), provides N bits of entropy.
+            return len(self.safe_summary(encoded)['salt']) < self.salt_entropy / math.log(62, 2)
```

Comment:
I'm starting to think if this is an expected change :thinking: This can force updates each time when the password is checked for password created with a `salt` passed manually to the `make_password()`.

Reply:
Good question. It will ensure that salt_entropy is met… which brings me back to the comment mentioned in the ticket: Does it make sense to support a `salt` for encode at all? You need to know the hasher to be able to provide the proper `salt` (think the bcrypt format which encodes rounds in it).

Sadly this will be a rather tricky issue to work on. I also think that `must_update` shouldn't do this calculation of the `salt_len` but rather the hashers should return it from `decode` so they are free to store it in a different form…

Reply:
If the hashers should each be making their own determination then should I open another PR to move `must_update` to be a required function for each hasher?

Reply:
Mhm that is what I was trying to avoid, because for most hasher a salt length is just that and `must_update` should easily be able to handle that globally if it is returned from `decode`. 

What this PR certainly misses (and what will show you the existing problems) is a test for the behavior of the `bcrypt` hasher. I think now it's `must_update` will *always* return `True` and set a salt *every* time.

Reply:
I'm not sure I see the issue. `BCryptSHA256PasswordHasher` defines its own `must_update` as
```
    def must_update(self, encoded):
        decoded = self.decode(encoded)
        return decoded['work_factor'] != self.rounds
```
and `BCryptPasswordHasher` inherits from it. Maybe bcrypt should take the salt into account in that function, but I don't see how it would always return true. I've added some tests for this as well. 

That said, I've noticed that most of the hashers have a custom `must_update` which does not account for the salt size. I think they should and if you agree then I'll add logic to them to check.

Reply:
@apollo13 Where do you stand on the `must_update` change? Is that something that should be addressed later? If so what would be the proper channel in which to raise this topic for discussion?

Reply:
@apollo13 Can we circle back to this? I've added a salt strength check to the pbkdf2 hasher as well as two tests related to checking the behavior of `must_update`. Maybe pbkdf2 is the only hasher that it makes sense to add this check to.

Reply:
> Maybe pbkdf2 is the only hasher that it makes sense to add this check to.

Why? It certainly also makes sense for Argon and the standard salted sha/md5 stuff. The main problem I see currently is that https://github.com/django/django/blob/d618712a3b9df95509a053cf9fa444f47d4d0c79/django/contrib/auth/hashers.py#L236 is simply wrong for some hashers, so yeah maybe we cannot provide a generic `must_update` method (ie the crypt hasher has a salt length of 2 [literally]).

> That said, I've noticed that most of the hashers have a custom must_update which do not account for the salt size. I think they should account for salt size and if you agree then I'll add logic to them to check.

Yes my initial idea was that all hasher you call `must_update` that would allow them to determine if salt changes are needed. Sadly I do not think it is possible to write a `salt` check in a generic way.

To be honest I do not have good ideas about that. For this to be mergable we'd at least have to have it working for Argon & pbkdf2 (sha/md5 is easy then) without to many hacks / code smells.

Reply:
Also what happens if one passes in a salt manually? ie literally `hasher.encode(self, password, salt="123",…)` how can we ever enforce an entropy there?

Reply:
> Yes my initial idea was that all hasher you call must_update that would allow them to determine if salt changes are needed. Sadly I do not think it is possible to write a salt check in a generic way.

> To be honest I do not have good ideas about that. For this to be mergable we'd at least have to have it working for Argon & pbkdf2 (sha/md5 is easy then) without to many hacks / code smells.

I agree. I think this will need to be hasher specific though the base password hasher can provide a default case. I hadn't considered manual salt overrides and that's a good question. Generally I think we need to have memory of a manual override which we do not have today, so we either add memory to the stored hash somehow or we require the caller to remember and allow them to override the entropy value used in `must_update`. I don't think either solution is very clean.

As for the hashers; the salted md5/sha hashers don't implement their own `must_update` so they're already covered by the base password hasher implementation. I've added a salt check to argon2 as well with a minor simplification that it does not check for no salt given that the rest of the check also assumes a salt exists.

Reply:
> Yes my initial idea was that all hasher you call must_update that would allow them to determine if salt changes are needed. Sadly I do not think it is possible to write a salt check in a generic way.

> To be honest I do not have good ideas about that. For this to be mergable we'd at least have to have it working for Argon & pbkdf2 (sha/md5 is easy then) without to many hacks / code smells.

I agree. I think this will need to be hasher specific. I hadn't considered manual salt overrides and that's a good question. Generally I think we need to have memory of a manual override which we do not have today, so we either add memory to the stored hash somehow or we require the caller to remember and allow them to override the entropy value used in `must_update`. I don't think either solution is very clean.

As for the hashers; the salted md5/sha hashers don't implement their own `must_update` so they're already covered by the base password hasher implementation. I've added a salt check to md5/sha1 and argon2. Argon2 has a minor simplification that it does not check for the existence of the salt given that the rest of the check also assumes a salt exists. For md5 and sha1 this add `must_update` to each.

Reply:
@apollo13 any thoughts on the manual salting question? Also I've squashed a few of my commits down to clean up the salt strength checking portion of this PR.

Reply:
@apollo13 Alternately, how do you feel about breaking out the update portion of this PR? I can make a separate PR to handle the discussion of updating and reduce this PR to change the measurement of the salt from characters to bits and to set the default to 128 bits.